[Hack Week] ApiFaker: match endpoints using query parameters

hichamboushaba opened the pull request:
### Description
As part of my Hack Week, I'm bringing back the ApiFaker project paqN3M-SI-p2, and I'm splitting the feature into multiple PRs.

This PR adds support for matching endpoints using query parameters too, in the endpoint definition, we can add a list of query parameters that will filter which endpoints to mock, the logic adapts for the syntax of Jetpack tunneled requests and regular endpoints.

### Steps to reproduce
1. Use debug build.
2. Open the app settings.
3. Tap on Developer options.
4. Tap on ApiFaker.
5. Add a new endpoint definition.
6. Enter the path and choose the endpoint type.
7. Tap on ➕ button to add some query parameters.

### Testing information
- Confirm that only endpoints matching the query parameters are matched, you can use the endpoint `/wc/v3/orders` with query parameter `search={some term}` and use the search feature of the order list to see if the logic works as expected.
- Preferrably test using WordPress.com login and site credentials login.

### The tests that have been performed
^

### Images/gif
<img width=360 src=https://github.com/user-attachments/assets/8a858bfe-dfb8-474e-8746-fa53c0fa9f76/>
<img width=360 src=https://github.com/user-attachments/assets/92827f66-9e22-4203-adad-80bdc8bea330/>

</br></br>
- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Add query parameters to the DB model

hichamboushaba committed:
Filter requests using query parameters

hichamboushaba committed:
Add a way to add query parameters in the UI

hichamboushaba committed:
Some fixes

hichamboushaba committed:
Log a warning message if multiple endpoints are matched

hichamboushaba committed:
Bump db version

hichamboushaba committed:
Handle query parameters extraction for Jetpack tunnel requests

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Add more tests

dangermattic commented:
<!--
  2 Errors: Please add tests for class <co...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please add tests for class <code>QueryParameterConverter</code> (or add <code>unit-tests-exemption</code> label to ignore this).</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

JorgeMucientes commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/EndpointProcessor.kt:
```diff
@@ -18,23 +20,33 @@ internal class EndpointProcessor @Inject constructor(
+            endpointDao.queryEndpoint(apiType, httpMethod, path.trimEnd('/'), body.orEmpty())
+        }.also {
+            if (it.size > 1) {
+                Log.w(
+                    LOG_TAG,
+                    "More than one endpoint matched the request: $request, " +
+                        "the endpoints matched are\n$it\n" +
+                        "The first one will be used."
+                )
+            }
```
Minor np. This log will be fired when adding two endpoints with the same path but different query parameters.

hichamboushaba replied:
Nice catch, I updated the logic in the last commit to avoid this.

JorgeMucientes reviewed (approved):
Nicely done @hichamboushaba. Code looks good and everything works as expected. 

I just left a really minor nitpick. But nothing concerning. Approving!

hichamboushaba committed:
Improve logic of warning log

This makes sure the log is not printed when the matched endpoints use different query parameters.
